src/hotspot/share/cds/aotCodeSource.hpp

```diff
@@ -0,0 +1,271 @@
+  FileType _file_type;
+  Group    _group;
+  int      _index; // index of this AOTCodeSource inside AOTCodeSourceConfig::_code_sources
+  time_t   _timestamp;
+  int64_t  _filesize;
+
+  static size_t header_size()      { return sizeof(AOTCodeSource); } // bytes
+  size_t path_offset()       const { return header_size(); }
+  size_t manifest_offset()   const { return path_offset() + _path_length + 1; }
+  static char* read_manifest(JavaThread* current, const char* path, size_t& manifest_length);
+
+public:
+  static AOTCodeSource* allocate(JavaThread* current, const char* path, int index, Group group,
+                                 bool from_cpattr = false, bool is_jrt = false);
+
+  size_t total_size()                const { return manifest_offset() + _manifest_length + 1; }
+  const char* path()                 const { return ((const char*)this) + path_offset();  }
+  size_t manifest_length()           const { return _manifest_length; }
+  const char* manifest()             const { return ((const char*)this) + manifest_offset(); }
+  bool must_exist()                  const { return _file_type != FileType::NOT_EXIST; }
+  bool must_not_exist()              const { return _file_type == FileType::NOT_EXIST; }
+  bool is_dir()                      const { return _file_type == FileType::DIR; }
+  int index()                        const { return _index; }
+  bool is_modules_image()            const { return _group == Group::MODULES_IMAGE; }
+  bool from_boot_classpath()         const { return _group == Group::BOOT_CLASSPATH; }
+  bool from_app_classpath()          const { return _group == Group::APP_CLASSPATH; }
+  bool from_module_path()            const { return _group == Group::MODULE_PATH; }
+  bool is_multi_release_jar()        const { return _is_multi_release_jar; }
+
+  // Only boot/app classpaths can contain unnamed module
+  bool has_unnamed_module()          const { return from_boot_classpath() || from_app_classpath(); }
+
+  char* get_cpattr() const;
+  AOTCodeSource* write_to_archive() const;
+
+  // Returns true IFF this AOTCodeSource is discovered from the -classpath or -Xbootclasspath/a by parsing the
+  // "Class-Path" attribute of a JAR file.
+  bool from_cpattr() const { return _from_cpattr; }
+  const char* file_type_string() const;
+  bool check(const char* runtime_path, bool has_aot_linked_classes) const;
+};
+
+// AOTCodeSourceConfig
+//
+// Keep track of the set of AOTCodeSources used when an AOTCache is created.
+// To load the AOTCache in a production run, the JVM must be using a compatible set of
+// AOTCodeSources (subject to checks in AOTCodeSourceConfig::validate(bool&).
+//
+// In general, validation is performed on AOTCodeSources to ensure the code sources used
+// during AOTCache creation is the same as when the AOTCache is used during runtime.
+// Non-existent entries are recored during AOTCache creation. Those non-existent entries
+// must not exist during runtime.
+//
```

Comment:
Typos:
- "subjected to AOTCodeSourceConfig::validate()" -- the function has two parameters, but we can omit them in this comment
- "validation is performed on *the* AOTCodeSources"
- "during AOTCache creation *are* the same"
- "on-existent entries are *recorded*"

Reply:
Fixed.